utils/githooks/pre-commit.d/40-gofmt.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# Runs pylint for the DAOS project as a commit hook.
+#
+# To get the most out of this hook the 'gh' command should be installed and working.
+
+set -ue
+
+if ! BRANCH=$(git rev-parse --abbrev-ref HEAD 2>/dev/null); then
+        echo "Failed to determine branch with git rev-parse"
+        exit 1
+fi
+
+# Try and use the gh command to work out the target branch, or if not installed
+# then assume master.
+if command -v gh > /dev/null 2>&1
+then
+        # If there is no PR created yet then do not check anything.
+        if ! TARGET=origin/$(gh pr view "$BRANCH" --json baseRefName -t "{{.baseRefName}}")
+        then
+                TARGET=HEAD
+        fi
+else
+        # With no 'gh' command installed then check against master.
+        echo "Install gh command to auto-detect target branch, assuming master."
+        TARGET=origin/master
+fi
+
+output=
+if [ $TARGET = "HEAD" ]
+then
+        echo "Checking against HEAD"
+        output=$(git diff HEAD^ --name-only | grep -e '.go$' | xargs gofmt -d)
+else
+        echo "Checking against branch ${TARGET}"
+        output=$(git diff $TARGET... --name-only | grep -e '.go$' | xargs gofmt -d)
+fi
+
+if [ -n "$output" ]; then
```

Comment:
We have other commit-hooks (including but not limited to the spelling one) that just auto-applies the fixes, although this is probably good enough.  The problem with auto-applying is you need to run "git add" afterwards in the commit hook for modified files and there's no easy way to know which they are.

Reply:
Yeah, most people touching .go files will have their editor set up to run gofmt on save anyhow.

Reply:
> > The problem with auto-applying is you need to run "git add" afterwards in the commit hook for modified files and there's no easy way to know which they are
> 
> Doesn't the copyright githook do exactly that?

https://github.com/daos-stack/daos/blob/master/utils/githooks/pre-commit.d/10-update-copyright#L63 Yes, although in that case it uses grep to know which files to run it on.  My point was we might not want to run it for every go file.